app/src/main/java/mozilla/lockbox/extensions/ViewNode+.kt

```diff
@@ -14,22 +16,37 @@ fun AssistStructure.ViewNode.dump(): String {
 private fun AssistStructure.ViewNode.dumpNode(sb: StringBuilder = StringBuilder()): StringBuilder {
     val name = htmlInfo?.tag ?: className?.split('.')?.last() ?: "unknown"
 
-    var attrs = listOf(
-        Pair("idEntry", idEntry ?: ""),
-        Pair("idPackage", idPackage ?: ""),
-        Pair("idType", idType ?: ""),
-        Pair("webDomain", webDomain ?: ""),
-        Pair("hint", hint ?: ""),
-        Pair("autofillValue", autofillValue?.textValue ?: ""),
-        Pair("autofillHints", autofillHints?.joinToString(", ") ?: ""),
-        Pair("autofillOptions", autofillOptions?.joinToString(", ") ?: "")
+    var attrs: List<Pair<String, String>> = listOf(
+        "idEntry" to idEntry,
+        "idPackage" to idPackage,
+        "idType" to idType,
+        "webDomain" to webDomain,
+        "hint" to hint,
+        "autofillValue" to
+            if (autofillValue != null && autofillValue?.isText!!) {
+                autofillValue!!.textValue.toString()
+            } else {
+                emptyString
+            },
+        "autofillHints" to
+            if (autofillHints.isNullOrEmpty()) {
+                emptyString
+            } else {
+                autofillHints!!.joinToString(", ")
+            },
+        "autofillOptions" to
+            if (autofillOptions.isNullOrEmpty()) {
+                emptyString
+            } else {
+                autofillOptions!!.joinToString(", ")
+            }
     )
 
     htmlInfo?.attributes?.let { attributes ->
-        attrs += attributes.map { Pair(it.first, it.second) }
+        attrs += attributes.map { it.first to it.second }
     }
 
-    attrs = attrs.filter { it.second != "" && it.second != "null" }
+    attrs = attrs.filter { it.first != emptyString && it.second != "null" }
```

Comment:
I think there's a misunderstanding of what this code is doing.

In your comment: 

```kotlin
autofillValue?.isText ?: autofillValue?.textValue ?: emptyString
```

`autofillValue?.isText ?` is a `Boolean?`. The `?:` elvis operator is a fallthrough if the left hand side is a `null`.

So, if `autofillValue` doesn't exist, then we try `autofillValue?.textValue`, which is of type `String?`, but we know that it's going to be `null`, so we get `emptyString`.

The interesting bit is what type this is going to be. The compiler thinks it's going to be a `Boolean || String` optional. Since kotlin doesn't have algebraic types, then the only thing it can conclude is that it's `Any?`.

The code you've committed:

```kotlin
            if (autofillValue != null && autofillValue?.isText!!) {
                autofillValue!!.textValue.toString()
            } else {
                emptyString
            },
```
works, but requires you force dereference (`!!`). Channeling Matz here, we should treat `!!` as syntactic vinegar: force-dereferencing should be avoided — because of the danger of `NullPointerException` — so they made it ugly to read.

Here's what I suggest:

```kotlin
    val viewAttrs = listOf(
        "idEntry" to idEntry,
        "idPackage" to idPackage,
        "idType" to idType,
        "webDomain" to webDomain,
        "hint" to hint,
        "autofillValue" to autofillValue?.let { if (it.isText) it.textValue.toString() else null },
        "autofillHints" to autofillHints?.joinToString(", "),
        "autofillOptions" to autofillOptions?.joinToString(", ")
    )

    val htmlAttrs = htmlInfo?.attributes?.let { attributes ->
        attributes.map { it.first to it.second }
    } ?: emptyList()

    val attrs = (viewAttrs + htmlAttrs).filter {
        it.second != null
    }
```

The `autofillValue`, `autofillHints` and `autofillOptions` all are nullable. `autofillValue` **must** be checked as to be of text type. If not, we emit `null`.

Then we filter out all the `null` valued pairs.

FWIW, I'm not especially happy we can't use 

```kotlin
val htmlAttrs = htmlInfo?.attributes?.toList() ?: emptyList()
```

but I'm not sure how to persuade kotlin that `Pair<String!, String!>` and `Pair<String!, String!>!` should be treated in the same way.